ta: pkcs11: fix EC generation and EC import

vesajaaskelainen opened the pull request:
The PKCS#11 standard does not allow one to have CKA_EC_POINT for private
keys but TEE internal API requires one to be present when performing
private key operations. Instead of calculating it each time it is needed
store it as hidden attribute.

This fixes EC private key generation to function as specified in standard.

There is backwards support for existing keys that has been created
inadvertently with CKA_EC_POINT included.

When importing EC private key also generate hidden EC public key for that.

Fixes:
- https://github.com/OP-TEE/optee_os/issues/5165

TODO:
- Add tests

Notes:
- It seems that Edwards curves support has the same problem with CKA_EC_POINT being allowed but that is not addressed in this PR.
---

Testing steps.

Initializing token:
```
root@qemuarm64-secureboot:~# export PKCS11_MODULE=/usr/lib/libckteec.so.0
root@qemuarm64-secureboot:~# export PKCS11_SLOT=0
root@qemuarm64-secureboot:~# export PKCS11_TOKEN=device
root@qemuarm64-secureboot:~# export PKCS11_SO_PIN=1234567890
root@qemuarm64-secureboot:~# export PKCS11_USER_PIN=1234
root@qemuarm64-secureboot:~# export PKCS11_EC_OBJECT_ID=223344
root@qemuarm64-secureboot:~# export PKCS11_EC_OBJECT_LABEL=myECImportedKey
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --slot-index ${PKCS11_SLOT} --init-token --label ${PKCS11_TOKEN} --so-pin ${PKCS11_SO_PIN}
Using slot with index 0 (0x0)
Token successfully initialized
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --slot-index ${PKCS11_SLOT} --init-pin --login --so-pin ${PKCS11_SO_PIN} --new-pin ${PKCS11_USER_PIN}
Using slot with index 0 (0x0)
User PIN successfully initialized
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --list-slots
Available slots:
Slot 0 (0x0): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token label        : device
  token manufacturer : Linaro
  token model        : OP-TEE TA
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.1
  serial num         : 0000000000000000
  pin min/max        : 4/128
Slot 1 (0x1): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
Slot 2 (0x2): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token ${PKCS11_TOKEN} --login --pin ${PKCS11_USER_PIN} --list-objects
```
Generating the EC key on the token and verifying that it works:
```
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token ${PKCS11_TOKEN} --login --pin ${PKCS11_USER_PIN} --keypairgen --key-type EC:prime256v1 --label myECkey
Key pair generated:
Private Key Object; EC
  label:      myECkey
  Usage:      sign, derive
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   0441045c36cc2124e38123289a2159b79b6198349dfbc25179f40168240525a79ad825678edab82c8e1d8aaedf858788a1269bb72fef818cb6ab849198045d76f34535
  EC_PARAMS:  06082a8648ce3d030107
  label:      myECkey
  Usage:      verify, derive
  Access:     local

# Note following _should_ work but this seems to be some incompatibility between openssl 3.0.9 and libp11#master.
root@qemuarm64-secureboot:~# openssl ec -engine pkcs11 -inform engine -in "pkcs11:token=${PKCS11_TOKEN};object=myECkey;type=public?pin-value=${PKCS11_USER_PIN}" -pubout -out /tmp/myECkey.pem
Engine "pkcs11" set.
read EC key
unable to enable public key encoding
205085A87F000000:error:030000A3:digital envelope routines:EVP_PKEY_set_params:invalid key:../openssl-3.0.9/crypto/evp/p_lib.c:2369:

# But we can workaround this by using pkcs11-tool and openssl conversion commands.
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token-label ${PKCS11_TOKEN} --pin ${PKCS11_USER_PIN} --login --read-object --type pubkey --label myECkey -o /tmp/myECkey.der
root@qemuarm64-secureboot:~# openssl ec -pubin -in /tmp/myECkey.der -inform DER -pubout -out /tmp/myECkey.pem
read EC key
writing EC key

root@qemuarm64-secureboot:~# echo "Hello World!" > /tmp/data.bin
root@qemuarm64-secureboot:~# openssl dgst -engine pkcs11 -keyform engine -sign "pkcs11:token=${PKCS11_TOKEN};object=myECkey;type=private?pin-value=${PKCS11_USER_PIN}" -out /tmp/data.bin.sig -sha256 /tmp/data.bin
Engine "pkcs11" set.
root@qemuarm64-secureboot:~# openssl dgst -engine pkcs11 -keyform engine -verify "pkcs11:token=${PKCS11_TOKEN};object=myECkey;type=public?pin-value=${PKCS11_USER_PIN}" -signature /tmp/data.bin.sig -sha256 /tmp/data.bin
Engine "pkcs11" set.
Verified OK
root@qemuarm64-secureboot:~# openssl dgst -verify /tmp/myECkey.pem -signature /tmp/data.bin.sig -sha256 /tmp/data.bin
Verified OK
```
Generating key externally and importing it and verifying that it works:
```
# Now this should also work but openssl 3.0.9 and libp11 are having a bad day
# If one removes the engine definition from openssl.cnf then it does work.
root@qemuarm64-secureboot:~# openssl ecparam -genkey -name prime256v1 -out /tmp/ec-private-key.pem
unable to write elliptic curve parameters
2020D6807F000000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):../openssl-3.0.9/crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?

# But we can workaround this by using pre-defined private key and converting it to needed formats
root@qemuarm64-secureboot:~# cat > /tmp/ec-private-key.pem << EOF
> -----BEGIN EC PARAMETERS-----
> BggqhkjOPQMBBw==
> -----END EC PARAMETERS-----
> -----BEGIN EC PRIVATE KEY-----
> MHcCAQEEIJU08dPkp+5YOZhoMqV2NN/8nTei1QD9pOTH2LCajsYxoAoGCCqGSM49
> AwEHoUQDQgAEP4JhSjgC8PyKkrqew5Mx/bNGWBGp2RpXh3gaWqFyPAVQp0mxTk5A
> 2DA0TWw2WvZ6UkjxSm12P9nxUWFEKXtR1w==
> -----END EC PRIVATE KEY-----
> EOF
root@qemuarm64-secureboot:~# openssl ec -in /tmp/ec-private-key.pem -out /tmp/ec-private-key.der
read EC key
writing EC key
root@qemuarm64-secureboot:~# openssl ec -in /tmp/ec-private-key.pem -pubout -outform DER -out /tmp/ec-public-key.der
read EC key
writing EC key
root@qemuarm64-secureboot:~# openssl ec -in /tmp/ec-private-key.pem -pubout -out /tmp/ec-public-key.pem
read EC key
writing EC key

# Importing private and public keys
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token ${PKCS11_TOKEN} --login --pin ${PKCS11_USER_PIN} --write-object /tmp/ec-private-key.der --type privkey --id ${PKCS11_EC_OBJECT_ID} --label ${PKCS11_EC_OBJECT_LABEL} --usage-sign --usage-derive
Created private key:
Private Key Object; EC
  label:      myECImportedKey
  ID:         223344
  Usage:      sign, derive
  Access:     sensitive
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token ${PKCS11_TOKEN} --login --pin ${PKCS11_USER_PIN} --write-object /tmp/ec-public-key.der --type pubkey --id ${PKCS11_EC_OBJECT_ID} --label ${PKCS11_EC_OBJECT_LABEL} --usage-sign --usage-derive
Created public key:
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   0441043f82614a3802f0fc8a92ba9ec39331fdb3465811a9d91a5787781a5aa1723c0550a749b14e4e40d830344d6c365af67a5248f14a6d763fd9f1516144297b51d7
  EC_PARAMS:  06082a8648ce3d030107
  label:      myECImportedKey
  ID:         223344
  Usage:      verify, derive
  Access:     none

# And testing it out
root@qemuarm64-secureboot:~# openssl dgst -engine pkcs11 -keyform engine -sign "pkcs11:token=${PKCS11_TOKEN};object=${PKCS11_EC_OBJECT_LABEL};type=private?pin-value=${PKCS11_USER_PIN}" -out /tmp/data.bin.sig -sha256 /tmp/data.bin
Engine "pkcs11" set.
root@qemuarm64-secureboot:~# openssl dgst -engine pkcs11 -keyform engine -verify "pkcs11:token=${PKCS11_TOKEN};object=${PKCS11_EC_OBJECT_LABEL};type=public?pin-value=${PKCS11_USER_PIN}" -signature /tmp/data.bin.sig -sha256 /tmp/data.bin
Engine "pkcs11" set.
Verified OK
root@qemuarm64-secureboot:~# openssl dgst -verify /tmp/ec-public-key.pem  -signature /tmp/data.bin.sig -sha256 /tmp/data.bin
Verified OK
```

embetrix commented:
@vesajaaskelainen I rebased your patch and apply it in on the imx-optee_os 3.19, I get a core dump when using the key :
```
# pkcs11-tool  --module /usr/lib/libckteec.so.0  --login  --write-object ec-private-key.der --type privkey --id 6232 --label ec-private-key --usage-sign --usage-derive
# OPENSSL_CONF=openssl.cnf openssl req -engine pkcs11 -new -key 6232 -keyform engine -out cert.pem -text -x509 -subj "/O=Embexus/CN=HSM-Test/emailAddress=info@embexus.com"
F/TC:? 0 trace_syscall:151 syscall #8 (syscall_check_access_rights)
F/TC:? 0 trace_syscall:151 syscall #50 (syscall_storage_obj_read)
F/TC:? 0 trace_syscall:151 syscall #24 (syscall_cryp_obj_get_info)
F/TC:? 0 trace_syscall:151 syscall #53 (syscall_storage_obj_seek)
F/TC:? 0 trace_syscall:151 syscall #28 (syscall_cryp_obj_close)
F/TC:? 0 plat_prng_add_jitter_entropy:72 0xF9
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_INIT rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS p#0 4@0x80018000, p#1 --- 0@0x0, p#2 out 4@0x80017000
D/TA:  entry_find_objects:697 PKCS11 session 1: finding objects
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS_FINAL p#0 4@0x80017000, p#1 --- 0@0x0, p#2 --- 0@0x0
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_FINAL rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 28@0x80018000, p#1 --- 0@0x0, p#2 out 20@0x80017000
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0x82/OBJECT_HANDLE_INVALID
D/TC:? 0 tee_ta_close_session:510 csess 0x560b0110 id 1
D/TC:? 0 tee_ta_close_session:529 Destroy session
D/TA:  close_ck_session:691 Close PKCS11 session 1
Segmentation fault (core dumped)

```

vesajaaskelainen commented:
Hmm...

```
root@qemuarm64-secureboot:~# openssl req -engine pkcs11 -new -key "pkcs11:token=${PKCS11_TOKEN};object=${PKCS11_EC_OBJECT_LABEL};type=private?pin-value=${PKCS11_USER_PIN}" -keyform engine -out /tmp/cert.pem -text -x509 -subj "/O=Embexus/CN=HSM-Test/emailAddress=info@embexus.com
"
Engine "pkcs11" set.
```

The OP-TEE traces:
```
I/TA: PKCS11 session 1: login
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_LOGIN rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_SESSION_INFO p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_ck_session_info:812 Get find on PKCS11 session 1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_SESSION_INFO rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS_INIT p#0 47@0x40018000, p#1 --- 0@0x0, p#2 --- 0@0x0
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_INIT rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 4@0x40018000
D/TA:  entry_find_objects:699 PKCS11 session 1: finding objects
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 28@0x40019000, p#1 --- 0@0x0, p#2 out 20@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 28@0x40019000, p#1 --- 0@0x0, p#2 out 20@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 279@0x40019000, p#1 --- 0@0x0, p#2 out 271@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 24@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 39@0x40019000, p#1 --- 0@0x0, p#2 out 31@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 25@0x40019000, p#1 --- 0@0x0, p#2 out 17@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 4@0x40018000
D/TA:  entry_find_objects:699 PKCS11 session 1: finding objects
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS_FINAL p#0 4@0x40018000, p#1 --- 0@0x0, p#2 --- 0@0x0
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_FINAL rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_SESSION_INFO p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_ck_session_info:812 Get find on PKCS11 session 1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_SESSION_INFO rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 24@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 34@0x40019000, p#1 --- 0@0x0, p#2 out 26@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 24@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0x12/ATTRIBUTE_TYPE_INVALID
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS_INIT p#0 35@0x40018000, p#1 --- 0@0x0, p#2 --- 0@0x0
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_INIT rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 4@0x40018000
D/TA:  entry_find_objects:699 PKCS11 session 1: finding objects
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_FIND_OBJECTS_FINAL p#0 4@0x40018000, p#1 --- 0@0x0, p#2 --- 0@0x0
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_FIND_OBJECTS_FINAL rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 28@0x40019000, p#1 --- 0@0x0, p#2 out 20@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 28@0x40019000, p#1 --- 0@0x0, p#2 out 20@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 279@0x40019000, p#1 --- 0@0x0, p#2 out 271@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 24@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 39@0x40019000, p#1 --- 0@0x0, p#2 out 31@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 24@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_GET_ATTRIBUTE_VALUE p#0 91@0x40019000, p#1 --- 0@0x0, p#2 out 83@0x40018000
D/TA:  entry_get_attribute_value:912 PKCS11 session 1: get attributes 0x2
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_SESSION_INFO p#0 4@0x40019000, p#1 --- 0@0x0, p#2 out 16@0x40018000
D/TA:  entry_ck_session_info:812 Get find on PKCS11 session 1
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_SESSION_INFO rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_SIGN_INIT p#0 16@0x40018000, p#1 --- 0@0x0, p#2 --- 0@0x0
F/TC:? 0 trace_syscall:150 syscall #8 (syscall_check_access_rights)
F/TC:? 0 trace_syscall:150 syscall #27 (syscall_cryp_obj_alloc)
F/TC:? 0 trace_syscall:150 syscall #15 (syscall_cryp_state_alloc)
F/TC:? 0 trace_syscall:150 syscall #24 (syscall_cryp_obj_get_info)
F/TC:? 0 trace_syscall:150 syscall #24 (syscall_cryp_obj_get_info)
F/TC:? 0 trace_syscall:150 syscall #29 (syscall_cryp_obj_reset)
F/TC:? 0 trace_syscall:150 syscall #24 (syscall_cryp_obj_get_info)
F/TC:? 0 trace_syscall:150 syscall #24 (syscall_cryp_obj_get_info)
F/TC:? 0 trace_syscall:150 syscall #31 (syscall_cryp_obj_copy)
D/TA:  entry_processing_init:672 PKCS11 session 1: init processing ECDSA SIGN
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_SIGN_INIT rc 0/OK
D/TA:  TA_InvokeCommandEntryPoint:143 PKCS11_CMD_SIGN_ONESHOT p#0 4@0x4001a000, p#1 in 32@0x40019298, p#2 out 72@0x40018990
F/TC:? 0 trace_syscall:150 syscall #8 (syscall_check_access_rights)
F/TC:? 0 trace_syscall:150 syscall #8 (syscall_check_access_rights)
F/TC:? 0 trace_syscall:150 syscall #39 (syscall_asymm_operate)
F/TC:? 0 trace_syscall:150 syscall #17 (syscall_cryp_state_free)
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_SIGN_ONESHOT rc 0/OK
F/TC:? 0 plat_prng_add_jitter_entropy:72 0xA2
D/TC:? 0 tee_ta_close_session:529 csess 0xc0e3e880 id 1
D/TC:? 0 tee_ta_close_session:548 Destroy session
D/TA:  close_ck_session:703 Close PKCS11 session 1
```

Now what I did in between was update to libp11's master to see if some openssl 3.x related issues would have been fixed.

There is this commit that might affect:
https://github.com/OpenSC/libp11/commit/b2ba7ceda3d516fa22a8ef2165d2db799e580695

Here is related test program for EC related crash:
https://github.com/OpenSC/libp11/commit/af8c0616b69d5ab23dcc4514042ada235f8b1738

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -5,6 +5,7 @@
 
 #include <assert.h>
 #include <inttypes.h>
+#include <mbedtls/asn1write.h>
```
It seems to be getting indirectly included by:
```c
#include <mbedtls/pk.h>
```
But lets add direct include for it.

embetrix replied:
@vesajaaskelainen without the include the build broke on 3.19

embetrix commented:
@vesajaaskelainen I update the libp11 to latest master  : 74497e0fa5b69b15790d6697e1ebce13af842d4c

it didn't solve the problem :-( 

# openssl version
OpenSSL 3.0.9 30 May 2023 (Library: OpenSSL 3.0.9 30 May 2023)

embetrix commented:
@vesajaaskelainen , I gave it a try using sign/verify with pkcs11-tool instead of openssl:

```
# openssl ecparam -genkey -name prime256v1 -out ec-private-key.pem
# openssl ec -in ec-private-key.pem -outform DER -out ec-private-key.der
read EC key
writing EC key
# openssl ec -in ec-private-key.pem -pubout -outform DER -out ec-public-key.der
read EC key
writing EC key

# pkcs11-tool  --module /usr/lib/libckteec.so.0  --login  --write-object ec-private-key.der --type privkey --id 6111 --label ec-private-key --usage-sign --usage-derive
Using slot 0 with a present tokenI/TA: PKCS11 session 1: login
 (0x0)
Created private key:
Private Key Object; EC
  label:      ec-private-key
  ID:         6111
  Usage:      sign, derive
  Access:     sensitive

# pkcs11-tool  --module /usr/lib/libckteec.so.0  --login  --write-object ec-public-key.der --type pubkey --id 6112 --label ec-public-key --usage-sign --usage-derive
Using slot 0 with a present token (0x0)
I/TA: PKCS11 session 1: login
Created public key:
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   044104bd78fbc63924ef76002f0e8e5337f8811ad30e911883715871b4fc5dd404d626a4a43bf9be3e0fd00a23b6c6504ceba99d20fe68ec7a1c5a77049a2877c11a43
  EC_PARAMS:  06082a8648ce3d030107
  label:      ec-public-key
  ID:         6112
  Usage:      verify, derive
  Access:     none

# pkcs11-tool --module /usr/lib/libckteec.so.0  --id 6111 -s -i data.bin -m ECDSA-SHA256 > data.sig
Using slot 0 with a present token (0x0)
I/TA: PKCS11 session 1: login
Using signature algorithm ECDSA-SHA256
# pkcs11-tool  --module /usr/lib/libckteec.so.0  --id 6112  --verify -m ECDSA-SHA256   --input-file data.bin --signature-file  data.sig
Using slot 0 with a present token (0x0)
Using signature algorithm ECDSA-SHA256
Signature is valid
```

there is something fishy with openssl or libp11

embetrix commented:
@vesajaaskelainen which opensc version are you using ?  I use yocto kirkstone and it has the opensc 0.22 version

vesajaaskelainen commented:
> @vesajaaskelainen which opensc version are you using ? I use yocto kirkstone and it has the opensc 0.22 version

On kirkstone.

```
PV="0.22.0"
SRCREV="c902e1992195e00ada12d71beb1029287cd72037"
```

vesajaaskelainen commented:
You can observe what is happening on Cryptoki level with pkcs11-spy:
```
export PKCS11SPY=/usr/lib/libckteec.so.0
export PKCS11SPY_OUTPUT=/tmp/pkcs11-spy.log
export PKCS11_MODULE=/usr/lib/pkcs11-spy.so
```

And then just look from the log.

I suppose `gdb` (or other debugger) is something that would aid in here figuring out what is crashing.

Something that jumps out from your log is:
```
D/TA:  TA_InvokeCommandEntryPoint:364 PKCS11_CMD_GET_ATTRIBUTE_VALUE rc 0x82/OBJECT_HANDLE_INVALID
```
Which is a bit weird error. With pkcs11-spy you can see what is going on there.

embetrix commented:
@vesajaaskelainen I figure out what my problem was : I imported the the private key and public key with different IDs !

specifying the same id solved my issue!

many thanks for your support, very appreciated  :-)

vesajaaskelainen commented:
> @vesajaaskelainen I figure out what my problem was : I imported the the private key and public key with different IDs !
> 
> specifying the same id solved my issue!
> 
> many thanks for your support, very appreciated :-)

:partying_face:

Thanks for testing it out!

Could you provide your Reviewed-by/Tested-by or such tag?

embetrix commented:
Reviewed-by/Tested-by @embetrix

vesajaaskelainen commented:
> Reviewed-by/Tested-by @embetrix

Oh... Could you give it in form like:
```
Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>
```

So we have proper "developer's certificate of origin", with real name and email, thanks.

I prefer to copy'n'paste it from here to make sure there are no mistakes and names are written properly.

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#using-reported-by-tested-by-reviewed-by-suggested-by-and-fixes

If you want to be left in anonymity that is ok too but then I will just omit the info in commit messages.

vesajaaskelainen commented:
> Reviewed-by: Ayoub Zaki [ayoub.zaki@embetrix.com](mailto:ayoub.zaki@embetrix.com) Tested-by: Ayoub Zaki [ayoub.zaki@embetrix.com](mailto:ayoub.zaki@embetrix.com)

Thanks! Applied tags.

Also added the missing include.

Rebased on top of master.

Re-tested.

vesajaaskelainen commented:
@etienne-lms would like to hear your input for this PR. Thanks!

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -857,6 +860,180 @@ static enum pkcs11_rc create_priv_key_attributes(struct obj_attrs **out,
+
+out:
+	TEE_Free(asnbuf);
+	asnbuf = NULL;
```
can remove `asnbuf = NULL;` and `buf = NULL;`

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1862,6 +1862,12 @@ check_parent_attrs_against_processing(enum pkcs11_mechanism_id proc_id,
+{
+	return (req_attr->id & PKCS11_CKA_OPTEE_FLAGS_HIDDEN) ==
+		PKCS11_CKA_OPTEE_FLAGS_HIDDEN;
+}
```
nitpicking: could be inlined.

etienne-lms reviewed (commented):
LGTM with minor comments.

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -1039,12 +1039,32 @@ struct pkcs11_attribute_head {
+ * bit 30 - PKCS11_CKF_ARRAY_ATTRIBUTE - works like in normal attributes
+ * bit 24-29 - reserved in case PKCS#11 standard starts to use them
+ * bit 16-24 - allocated for OP-TEE attribute flags
+ * bit 0-15 - allocated for attribute identifier
```
As far I can tell only the array attribute and vendor bits are from the standard. Even the array could be handled differently when vendor bit is set but I think it is good to use the same bits.

vesajaaskelainen commented:
> LGTM with minor comments.

Did the changes.

Re-based on top of master.

Re-tested.

jkroon81 commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -1039,12 +1039,32 @@ struct pkcs11_attribute_head {
+ * bit 31 - PKCS11_CKA_VENDOR_DEFINED
+ * bit 30 - PKCS11_CKF_ARRAY_ATTRIBUTE - works like in normal attributes
+ * bit 24-29 - reserved in case PKCS#11 standard starts to use them
+ * bit 16-24 - allocated for OP-TEE attribute flags
```
I think this should be "bit 16-23", not "16-24" ?

jkroon81 commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -857,6 +860,176 @@ static enum pkcs11_rc create_priv_key_attributes(struct obj_attrs **out,
+
+	if (get_attribute_ptr(temp, PKCS11_CKA_VALUE,
+			      &a_ptr, &a_size) || !a_ptr) {
+		EMSG("No EC_PARAMS attribute found in private key");
```
s/EC_PARAMS/VALUE ?

etienne-lms reviewed (commented):
Hello @vesajaaskelainen.

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"ta: pkcs11: introduce PKCS11_CKF_ARRAY_ATTRIBUTE",
"ta: pkcs11: specify OP-TEE's vendor specific attribute scheme".
"ta: pkcs11: add hidden EC point support"

For commit "ta: pkcs11: add OP-TEE hidden attribute extension".
The change looks good to me. Howverever, I think `attribute_is_exportable()` should also  return `false` when OPTEE_FLAGS_HIDDEN is set, for consistency, even if not strictly needed by its current caller `entry_get_attribute_value()`. What do you think?
Otherwise the change looks good to me.

For commit "ta: pkcs11: fix EC private key import":
The change does not cover Edwards curves. Could you either address that or write in the commit message that the change covers only supported ECC NIST curves?
Otherwise, the change looks good to me.

vesajaaskelainen commented:
> For commit "ta: pkcs11: fix EC private key import":
> The change does not cover Edwards curves. Could you either address that or write in the commit message that the change covers only supported ECC NIST curves?
> Otherwise, the change looks good to me.

Right. I did not have any tests for Edwards curves -- I had it as comment that this did not have that support.

But let me add own commit for that while working on here.

vesajaaskelainen commented:
Umh... How was the Edwards curve support tested out?

I just generated the Edwards key pair in the pkcs11 ta.:
```
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token ${PKCS11_TOKEN} --login --pin ${PKCS11_USER_PIN} --keypairgen --key-type EC:edwards25519 --label myEDkey
Key pair generated:
Private Key Object; EC_EDWARDS
  label:      myEDkey
  Usage:      sign, derive
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; EC_EDWARDS  EC_POINT 255 bits
  EC_POINT:   9d25ccca8fde4261f0bdf193077776cccf681a1a5fe8d8f35f1fbf8f7eaf4c18
  EC_PARAMS:  130c656477617264733235353139 (OID 2.21.100.119.97.114.100.115.50.53.53.49.57)
  label:      myEDkey
  Usage:      verify, derive
  Access:     local
```

Then tried to get out public key for it:
```
root@qemuarm64-secureboot:~# pkcs11-tool --module ${PKCS11_MODULE} --token-label ${PKCS11_TOKEN} --pin ${PKCS11_USER_PIN} --login --read-object --type pubkey --label myEDkey -o /tmp/myEDkey.der
error: cannot decode EC_POINT
Aborting.
```

Then looking at the pkcs11-spy capture:
```
13: C_GetAttributeValue
2023-09-04 16:39:10.289
[in] hSession = 0x1
[in] hObject = 0x1
[in] pTemplate[1]: 
    CKA_EC_PARAMS         000000555efa3d90 / 14
[out] pTemplate[1]: 
    CKA_EC_PARAMS         000000555efa3d90 / 14
    00000000  13 0C 65 64 77 61 72 64 73 32 35 35 31 39        ..edwards25519  
Returned:  0 CKR_OK
...
15: C_GetAttributeValue
2023-09-04 16:39:10.312
[in] hSession = 0x1
[in] hObject = 0x1
[in] pTemplate[1]: 
    CKA_EC_POINT          000000555efa3ee0 / 32
[out] pTemplate[1]: 
    CKA_EC_POINT          000000555efa3ee0 / 32
    00000000  9D 25 CC CA 8F DE 42 61 F0 BD F1 93 07 77 76 CC  .%....Ba.....wv.
    00000010  CF 68 1A 1A 5F E8 D8 F3 5F 1F BF 8F 7E AF 4C 18  .h.._..._...~.L.
Returned:  0 CKR_OK
```

There seems to be some expectation differences between pkcs11-tool and standard:

https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.1/csd01/pkcs11-spec-v3.1-csd01.html#_Toc98177266
https://datatracker.ietf.org/doc/html/rfc8032#section-7.1

Then observing what was done... we have 32 bytes for public key and the EC PARAMS seem to match too.

So to me it looks like the pkcs11 ta might be doing the right thing but tooling feels a bit different:
https://github.com/OpenSC/OpenSC/blame/master/src/tools/pkcs11-tool.c#L5806

Just updated opensc to latest release and and libp11 to master.

For libp11 there is a ticket to get support for Edwards curve so openssl engine path does not seem feasible.

I suppose I could next get latest pkcs11-provider compiled on my test bed.

etienne-lms commented:
Looking at pkcs11 spec [v2.40-errata1](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html), CKA_EC_POINT is DER encoded:
> 2.3.3 ECDSA public key objects
(...)
CKA_EC_POINT1,4 | Byte array | DER-encoding of ANSI X9.62 ECPoint value Q

In pkcs11 spec [v3.0](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html), both ECDSA and Edwards public key CKA_EC_POINT is DER encoded:
> 2.3.3 ECDSA public key objects
(...)
CKA_EC_POINT1,4 | Byte array | DER-encoding of ANSI X9.62 ECPoint value Q

> 2.3.5 Edwards Elliptic curve public key objects
(...)
CKA_EC_POINT1,4 | Byte array | DER-encoding of the b-bit public key value in little endian order as defined in RFC 8032

In spec v3.1 (you pointed), CKA_EC_POINT is a byte series:
> 6.3.3 Short Weierstrass Elliptic Curve public key objects
(...)
CKA_EC_POINT1,4 | Byte array | Public key bytes in little endian order as defined in RFC 8032

vesajaaskelainen commented:
> Looking at pkcs11 spec [v2.40-errata1](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html), CKA_EC_POINT is DER encoded:
...
> 
> In pkcs11 spec [v3.0](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html), both ECDSA and Edwards public key CKA_EC_POINT is DER encoded:
...
> In spec v3.1 (you pointed), CKA_EC_POINT is a byte series:

Sigh... I sent question to pkcs#11 commitee's feedback channel:
https://lists.oasis-open.org/archives/pkcs11-comment/202309/msg00000.html

vesajaaskelainen commented:
Edwards curve support would need a bit of extra work to make it to make it more usable. We ware missing different forms of Edwards curves and then this storage would need to be sorted out.

"The mechanism can only generate EC public/private key pairs over the curves edwards25519 and edwards448 as defined in RFC 8032 or the curves id-Ed25519 and id-Ed448 as defined in RFC 8410"

So I am wondering should I just concentrate on classical elliptic curves to get those fixed and then after that let's fix the Edwards curve support as that would require quite a bit of testing and verification work.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -857,6 +860,176 @@ static enum pkcs11_rc create_priv_key_attributes(struct obj_attrs **out,
+		break;
+	case TEE_ECC_CURVE_NIST_P521:
+		ec_curve = MBEDTLS_ECP_DP_SECP521R1;
+		break;
```
I would leave this for other PR to fix Edwards curves support.

vesajaaskelainen commented:
Thanks for the review so far!

Applied given tags so far.

Did the suggested changes.

Rebased on master.

Re-tested.

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.h:
```diff
@@ -162,6 +163,12 @@ check_mechanism_against_processing(struct pkcs11_session *session,
 
+static inline bool attribute_is_hidden(struct pkcs11_attribute_head *req_attr)
+{
+	return (req_attr->id & PKCS11_CKA_OPTEE_FLAGS_HIDDEN) ==
```
`(req_attr->id & PKCS11_CKA_OPTEE_FLAGS_MASK) == PKCS11_CKA_OPTEE_FLAGS_HIDDEN`

vesajaaskelainen replied:
If we would introduce new flags then this would start to fail.

etienne-lms replied:
Sorry, indeed `PKCS11_CKA_OPTEE_FLAGS_HIDDEN` already consider `PKCS11_CKA_VENDOR_DEFINED` flags. Discard my comment, this looks good to me.

etienne-lms reviewed (commented):
For commit "ta: pkcs11: add OP-TEE hidden attribute extension": mask to fix,
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` once addressed.

For commit "ta: pkcs11: fix EC private key import":
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>`.

vesajaaskelainen commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -1186,7 +1186,19 @@ enum pkcs11_attr_id {
+	 * should not include CKA_EC_POINT include hidden one so that it does
+	 * not need to be calculated on each operation.
 	 */
+	PKCS11_CKA_OPTEE_HIDDEN_EC_POINT = PKCS11_CKA_VENDOR_DEFINED |
```
Ok

etienne-lms reviewed (commented):
For commit "ta: pkcs11: add OP-TEE hidden attribute extension"
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com> as-is.

jforissier commented:
Thanks @vesajaaskelainen you're doing a great job with the PKCS#11 TA :+1: much appreciated
